pkg/validators/am0010_k8s_resource_names.go

```diff
@@ -0,0 +1,252 @@
+
+func validateLabel(mb types.MetaBundle) (failures []failureMsg) {
+	if msg := isValidLabelName(mb.AddonMeta.Label); !msg.IsEmpty() {
+		failures = append(failures, prefixedFailureMsg("label", msg))
+	}
+
+	return failures
+}
+
+func validateTargetNamespace(mb types.MetaBundle) (failures []failureMsg) {
+	if msg := isValidNamespaceName(mb.AddonMeta.TargetNamespace); !msg.IsEmpty() {
+		failures = append(failures, prefixedFailureMsg("targetNamespace", msg))
+	}
+
+	return failures
+}
+
+func validateAllNamespaces(mb types.MetaBundle) []failureMsg {
+	var result []failureMsg
+
+	if msgs := areValidNamespaceNames(mb.AddonMeta.Namespaces...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaces", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidAnnotationNames(keys(mb.AddonMeta.NamespaceAnnotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidLabelNames(keys(mb.AddonMeta.NamespaceLabels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonAnnotations(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonAnnotations == nil {
+		return []failureMsg{}
+	}
+
+	annotations := *mb.AddonMeta.CommonAnnotations
+
+	result := make([]failureMsg, 0, len(annotations))
+
+	if msgs := areValidAnnotationNames(keys(annotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonLabels(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonLabels == nil {
+		return []failureMsg{}
+	}
+
+	labels := *mb.AddonMeta.CommonLabels
+
+	result := make([]failureMsg, 0, len(labels))
+
+	if msgs := areValidLabelNames(keys(labels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+type k8sName string
+
+const (
+	k8sNameAnnotation = k8sName("annotation")
+	k8sNameLabel      = k8sName("label")
+	k8sNameNamespace  = k8sName("namespace")
+)
+
+func areValidAnnotationNames(names ...string) []failureMsg {
+	return areValidNames(k8sNameAnnotation, names...)
+}
+
+func areValidLabelNames(names ...string) []failureMsg {
+	return areValidNames(k8sNameLabel, names...)
+}
+
+func areValidNamespaceNames(names ...string) []failureMsg {
+	return areValidNames(k8sNameNamespace, names...)
+}
+
+func areValidNames(nameType k8sName, names ...string) []failureMsg {
+	validationFunc := k8sNameToValidationFunc(nameType)
+
+	msgs := make([]failureMsg, 0, len(names))
+
+	for _, name := range names {
+		if msg := validationFunc(name); !msg.IsEmpty() {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs
+}
+
+func k8sNameToValidationFunc(nameType k8sName) func(string) failureMsg {
```

Comment:
lgtm

Thanks for refactoring. I would say this is overkill for my personal taste, and adds extra cognitive complexity where you have to jump around some more to understand the code, but I don't have strong feelings about this.

Good job on your first PR! Merge it when you are ready. :rocket:

Reply:
It should settle out once helpers are repacked, thanks.